x/wasm/client/cli/tx.go

```diff
@@ -420,18 +440,84 @@ func GrantAuthorizationCmd() *cobra.Command {
 			if exp == 0 {
 				return errors.New("expiration must be set")
 			}
-			_ = clientCtx
-			_ = grantee
-			_ = msgs
-			_ = once
-			_ = contract
 
-			return errors.New("not implemented")
+			allowAllMsgs, err := cmd.Flags().GetBool(flagAllowAllMsgs)
+			if err != nil {
+				return err
+			}
+
+			noTokenTransfer, err := cmd.Flags().GetBool(flagNoTokenTransfer)
+			if err != nil {
+				return err
+			}
+
+			var limit types.ContractAuthzLimitX
+			switch {
+			case maxFundsStr != "" && maxCalls != 0 && !noTokenTransfer:
+				maxFunds, err := sdk.ParseCoinsNormalized(maxFundsStr)
+				if err != nil {
+					return fmt.Errorf("max funds: %s", err)
+				}
+				limit = types.NewCombinedLimit(maxCalls, maxFunds...)
+			case maxFundsStr != "" && !noTokenTransfer:
+				maxFunds, err := sdk.ParseCoinsNormalized(maxFundsStr)
+				if err != nil {
+					return fmt.Errorf("max funds: %s", err)
+				}
+				limit = types.NewMaxFundsLimit(maxFunds...)
+			case maxCalls != 0 && noTokenTransfer:
+				limit = types.NewMaxCallsLimit(maxCalls)
+			default:
+				limit = types.UndefinedLimit{}
```

Comment:
you can fail fast here already
```suggestion
				limit = errors.New("invalid limit setup")
```